src/Sparrow.Server/AsyncManualResetEvent.cs

```diff
@@ -93,29 +93,37 @@ public Task<bool> WaitAsync()
                 return _tcs.Task;
             }
 
-            public IDisposable Register(CancellationToken token)
+            [Pure]
+            public async Task<bool> WaitAsync(CancellationToken token)
             {
                 var tcs = _tcs;
-                return token.Register(() => tcs.TrySetCanceled());
+                await using (token.Register(() => tcs.TrySetCanceled()))
```

Comment:
hmm when thinking about this, I'm not sure you can simply cancel the `_tcs`, others can await it... 
consider you have 2 threads doing `WaitAsync` with different token, so one will cancel the other.
you need to do `_parent.WaitAsync(token)`

Reply:
Done, used "WithCancellation" method instead to not cancel the task, only throw and stop waiting in case the token was canceled